### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by jetSett**
Note : we still must write some tests for auth.

I already tested at hand user creation+login,it seems to work as espected but Pierre will edit tests

### CODE REVIEW

**Comment by poechsel** [comment]
I would prefer to wait for the test to be written (or at least working) before merging.

File: tozti/core_schemas.py [hunk 1]
```diff
@@ -15,6 +15,21 @@
 }
 
 
+"""
+Hash of the user's password (SHA256)
+The hash is represented by a base10 integer in big endian
```
Comment (jetSett) [inline tozti/core_schemas.py]: not true anymore, I edit that

File: tozti/auth/utils.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        mac.add_first_party_caveat(json.dumps(a))
+            
+    for key, value in kwargs.items():
+        mac.add_first_party_caveat('{}={}'.format(json.dumps(key), json.dumps(value)))
+    
+    return mac
+
+class BadPasswordError(tozti.utils.APIError):
+    def __init__(self):
+        super().__init__(msg=self.detail)
```
Comment (Lapin0t) [inline tozti/auth/utils.py]: I'm not so fond of that, what did you want to achieve? tozti.utils.APIError should already provide a sufficient constructor. Maybe you want to change APIError to something cleaner (i would actually change the constructor to [init attributes...] super().__init__(self.to_response())). If you do that make sure it doesn't break tozti.app.error_handler.

File: tozti/__main__.py [hunk 3]
```diff
@@ -80,6 +80,40 @@ def find_exts():
...
+    config = {}
+    required = {"http":["host", "port"], 
+                "mongodb":["host", "port"], 
+                "cookie": ["private_key", "public_key"]}
+    with open(path) as s:
+        config = toml.load(s)
+
+    for major in required:
+        if not major in config:
+            raise Exception("Expected a {} entry".format(major))
```
Comment (Lapin0t) [inline tozti/__main__.py]: It's not good style to raise `Exception`, either raise `ValueError` or create some `ConfigError` to have meaningful exceptions.

Comment (jetSett) [reply tozti/__main__.py]: @poechsel this is your code, can you fix that ?

**Comment by Lapin0t** [comment]
Cool, good code quality 👍.

On L81-L96 of `tozti/auth/__init__.py` there is some code duplication (you are stealing some job of `Store.create` which validates the user input properly) but i'm gonna refactor `Store` anyway so it can be left like that for now, it's gonna be cleaner in a moment. Still it would be good to wrap the whole content in a `"data"` property (ie access `data["data"][{foo}]`, not `data[{foo}]`) to be consistent with JSON-API.

Stil i'm not sure it really makes sense to store the `(login, hash)` data as a resource: it's really something purely internal, we don't want to give it an UUID etc. It may be better to store that in another collection of the database. I'll take care of this.

**Comment by Lapin0t** [comment]
So i changed the DB, as far as i'm concerned it's in a mergeable state now. There are some stuff to refine but it can come later.

File: tozti/auth/__init__.py [hunk 4]
```diff
@@ -52,13 +52,10 @@
     except IndexError:
         raise BadJsonError()
 
-    hash_string = await req.app['tozti-store'].hash_by_login(login)
-    print(hash_string)
-    hash = str.encode(hash_string)
```
Comment (jetSett) [inline tozti/auth/__init__.py]: Sorry for debug code still in the commit T_T

**Comment by jetSett** [comment]
Just going to understand the code, change some errors (for example in storage.hash_by login) and I'll merge that in the evening

**Pull request merged**